tools/ci/format_changelog

```diff
@@ -0,0 +1,164 @@
+EOF
+  else
+    read -d '' pr_body <<EOF
+${pr_header}
+${changelog_content}
+EOF
+  fi
+  # for multi-line output
+  echo "pr_body<<EOF" >> $GITHUB_OUTPUT
+  echo "${pr_body}" >> $GITHUB_OUTPUT
+  echo "EOF" >> $GITHUB_OUTPUT
+}
+
+set_new_changelog_content() {
+  read -d '' new_changelog <<EOF
+${changelog_content}
+${current_changelog}
+EOF
+
+echo "$new_changelog" > CHANGELOG.md
+}
+
+# checks for tags in each changeset entry and append to tags.json
+match_tags() {
+  changesets_with_index=$(jq -r 'to_entries | .[] | "\(.key) \(.value | join(" "))"' changesets.json)
+
+  echo "$changesets_with_index" | while IFS= read -r line; do
+    index="${line%% *}"
+    changeset_content="${line#* }"
+    changeset_formatted=$(jq -r --argjson idx "$index" '.[$idx] | join("\n")' changesets.json)
+    found_tag=""
+    for tag in "${tags_list[@]}"; do
+      if [[ "$changeset_content" =~ $tag ]]; then
+        found_tag=${tag:1}
+        jq --arg key "$found_tag" --arg val "$changeset_formatted" \
+         '.[$key] += [$val]' tags.json > tmp.json && mv tmp.json tags.json
+      fi
+    done
+    if [[ $found_tag == "" ]] && [[ ! -z $changeset_content ]]; then
+      found_tag="untagged"
+      jq --arg key "$found_tag" --arg val "$changeset_formatted" \
+       '.[$key] += [$val]' tags.json > tmp.json && mv tmp.json tags.json
+    fi
+  done
+}
+
+cleanup() {
+  rm -f CHANGELOG.md.tmp
+  rm -f changesets.json
+  rm -f tags.json
+}
+
+### SCRIPT STARTS HERE ###
+cp CHANGELOG.md CHANGELOG.md.tmp
+tail -n +2 CHANGELOG.md > CHANGELOG.md.tmp
+
+pnpm changeset version
+
+version=$(jq -r '.version' package.json)
+echo "version=$version" >> $GITHUB_OUTPUT
+
+read -d '' changelog_content <<EOF
+# Changelog Chainlink Core
+
+## ${version} - PREVIEW
+
+EOF
+
+current_changelog=$(cat CHANGELOG.md.tmp)
+is_current_version=false
+checking_current_changeset=false
+changesets_index=0
+tags_list=( "#nops" "#added" "#changed" "#removed" "#updated" "#deprecation_notice" "#breaking_change" "#db_update" "#wip" "#bugfix" "#internal" "#untagged")
+
+create_changesets_json
+create_tags_json
+
+while IFS= read -r line; do
+  # break when hits the next version
+  if [[ $line == "## "* ]] && [[ $is_current_version = true ]]; then
+    break
+  fi
+
+  # look for the latest version
+  if [[ $line == "## "* ]]; then
+    is_current_version=true
+  fi
+
+  if [[ $is_current_version = true ]]; then
+    # saving each changeset to changeset.json
+    # check for start of changeset entry as it could be multi-lined entry
+    if [[ $line == "- ["* ]] && [[ $checking_current_changeset = true ]]; then
+      checking_current_changeset=false
+      changesets_index=$((changesets_index+1))
+      append_changeset_content "$line"
+    elif [[ $line == "- ["* ]] && [[ $checking_current_changeset = false ]]; then
+      checking_current_changeset=true
+      changesets_index=$((changesets_index+1))
+      append_changeset_content "$line"
```

Comment:
What does `checking_current_changeset` do here?

Seems like it just switches between `true` and `false`, and the statements within the two ifs are the same otherwise.

Reply:
this is related to whether we are checking for a current changeset entry or not since sometimes the dev might have a multi-line text for a changeset entry.

Reply:
hmm which I think you're getting at is  this `changesets_index=$((changesets_index+1))` exists in both conditions. let me check

Reply:
ahh yeah - great catch! it was unneeded :)